paddle/fluid/framework/scope.h

```diff
@@ -124,7 +141,8 @@ class Scope {
   DISABLE_COPY_AND_ASSIGN(Scope);
 
  private:
-  mutable std::mutex mutex_;
+  mutable RWLock kids_lock_;
+  mutable RWLock vars_lock_;
```

Comment:
for discussion: We might need to have a new class `ThreadsafeScope` inherit from `Scope`. In cases when locking is not needed, they can use the parent class.

Reply:
```ON_INFER``` macro free all lock and lock is needed now... because a few ops use Scope::Var and Scope::Rename to rename it...

Reply:
So a base class can be used for all cases like inference with no locks, and child class can be used in threaded executors.

Reply:
Oh, can do it in a single pull request though

Reply:
Good idea, I will fix it up in follow up PRs